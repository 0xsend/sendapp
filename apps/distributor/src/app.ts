import express, { type Request, type Response, Router } from 'express'
import pino from 'pino'
import { StandardMerkleTree } from '@openzeppelin/merkle-tree'
import { selectAll } from 'app/utils/supabase/selectAll'
import { createSupabaseAdminClient } from './supabase'
import { DistributorV2Worker } from './distributorv2'

const logger = pino({
  level: process.env.LOG_LEVEL || 'info',
  // ...other Pino options
})

// Initialize DistributorWorker
const distributorV2Worker = new DistributorV2Worker(logger)

// Initialize Express app
const app = express()

app.use(express.json())

// Example of a route (You'll need to define your routes as per your application's need)
app.get('/', (req, res) => {
  res.send({ root: true })
})

const distributorRouter = Router()

distributorRouter.get('/v2', async (req: Request, res: Response) => {
  res.json({
    distributor: true,
    ...distributorV2Worker.toJSON(),
  })
})

// Middleware for checking authorization
const checkAuthorization = (req: Request, res: Response, next: () => void) => {
  if (!req.headers.authorization?.includes(process.env.SUPABASE_SERVICE_ROLE as string)) {
    res.status(401).json('Unauthorized')
    return
  }
  next()
}

distributorRouter.post('/merkle', checkAuthorization, async (req: Request, res: Response) => {
  const { id } = req.body as { id: string }

  if (!id) {
    res.status(400).json('Missing id')
    return
  }

  logger.info({ id }, 'Received request to distribution merkle root')

  const supabaseAdmin = createSupabaseAdminClient()
  const { data: shares, error: sharesError } = await selectAll(
    supabaseAdmin
      .from('distribution_shares')
      .select('index, address, amount_after_slash::text', { count: 'exact' })
      .eq('distribution_id', id)
      .order('index', { ascending: true })
  )
  if (sharesError) {
    logger.error(sharesError, 'Error fetching distribution shares')
    res.status(500).json({ error: 'Error fetching distribution shares' })
    return
  }

  if (shares === null || shares.length === 0) {
    logger.error('No shares found for distribution', { id })
    res.status(500).json({ error: 'No shares found for distribution' })
    return
  }
  try {
    logger.info(`Found ${shares.length} shares for distribution ${id}`)

    const tree = StandardMerkleTree.of(
      shares.map(({ index, address, amount_after_slash }, i) => [
        index,
        address,
        BigInt(amount_after_slash),
      ]),
      ['uint256', 'address', 'uint256']
    )

    // this is what the user will need to submit to claim their tokens
    const proofs: string[][] = []
    for (const [i] of tree.entries()) {
      const proof = tree.getProof(i)
      proofs[i] = proof
    }

    const total = shares.reduce(
      (acc, { amount_after_slash }) => acc + BigInt(amount_after_slash),
      0n
    )
    const result = {
      id,
      root: tree.root,
      total: total.toString(),
      proofs,
      shares,
      tree: tree.dump(),
    }

    const response = JSON.parse(
      JSON.stringify(result, (key, value) => (typeof value === 'bigint' ? value.toString() : value))
    )

    res.json(response)
  } catch (err) {
    logger.error(err, 'Error while fetching shares')
    res.status(500).json({ error: 'Error while fetching shares' })
  }
})

distributorRouter.post('/v2', checkAuthorization, async (req, res) => {
  const { id } = req.body as { id: string }
  logger.info({ id }, 'Received request to calculate distribution')
  try {
    await distributorV2Worker.calculateDistribution(id)
  } catch (err) {
    logger.error(err, 'Error while calculating distribution')
    throw err
  }

  res.json({
    distributor: true,
    id: id,
  })
})

app.use('/distributor', distributorRouter)

export default app
export { app }
