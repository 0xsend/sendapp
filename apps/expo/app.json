{
  "expo": {
    "name": "send",
    "slug": "send",
    "scheme": "send",
    "version": "1.0.0",
    "orientation": "portrait",
    "icon": "./assets/icon.png",
    "userInterfaceStyle": "automatic",
    "splash": {
      "image": "./assets/splash.png",
      "resizeMode": "contain",
      "backgroundColor": "#ffffff"
    },
    "updates": {
      "fallbackToCacheTimeout": 0
    },
    "assetBundlePatterns": ["**/*"],
    "ios": {
      "supportsTablet": true,
      "bundleIdentifier": "send.app"
    },
    "android": {
      "adaptiveIcon": {
        "foregroundImage": "./assets/adaptive-icon.png",
        "backgroundColor": "#FFFFFF"
      },
      "package": "send.app"
    },
    "web": {
      "favicon": "./assets/favicon.png"
    },
    "plugins": [
      [
        "expo-image-picker",
        {
          "photosPermission": "The app accesses your photos to let you share them with your friends."
        }
      ],
      "expo-router"
    ]
  }
}
