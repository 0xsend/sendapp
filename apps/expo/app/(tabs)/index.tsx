import { Container, ScrollView } from '@my/ui'
import { HomeScreen } from 'app/features/home/screen'
import { Stack } from 'expo-router'

export default function HomeTabScreen() {
  return (
    <>
      <Stack.Screen
        options={{
          title: 'Send',
          headerShown: false, // We'll use the header from the tabs layout
        }}
      />

      <Container
        safeAreaProps={{
          edges: ['left', 'right'],
          style: { flex: 1 },
        }}
        flex={1}
        backgroundColor="$background"
      >
        <ScrollView
          flex={1}
          contentContainerStyle={{
            flexGrow: 1,
            paddingTop: 10,
            paddingBottom: 20,
          }}
          showsVerticalScrollIndicator={false}
          bounces={true}
          overScrollMode="always" // Android scroll indicator
        >
          <HomeScreen />
        </ScrollView>
      </Container>
    </>
  )
}
