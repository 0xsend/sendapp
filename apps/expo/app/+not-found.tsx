import { Stack } from 'expo-router/build/layouts/Stack'
import { ScreenContainer } from 'apps-expo/components/layout/ScreenContainer'
import { Paragraph } from '@my/ui'

export default function NotFoundScreen() {
  return (
    <>
      <Stack.Screen
        options={{
          title: 'Not Found',
        }}
      />
      <ScreenContainer>
        <Paragraph>NotFoundScreen</Paragraph>
      </ScreenContainer>
    </>
  )
}
