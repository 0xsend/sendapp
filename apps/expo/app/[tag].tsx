import { Stack as StackRouter } from 'expo-router/build/layouts/Stack'
import { ScreenContainer } from 'apps-expo/components/layout/ScreenContainer'
import { Paragraph } from '@my/ui'

export default function TagScreenScreen() {
  return (
    <>
      <StackRouter.Screen
        options={{
          title: 'History',
        }}
      />
      <ScreenContainer>
        <Paragraph>TagScreenScreen</Paragraph>
      </ScreenContainer>
    </>
  )
}
