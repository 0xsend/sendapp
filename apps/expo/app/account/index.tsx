import { ScreenContainer } from 'apps-expo/components/layout/ScreenContainer'
import { Stack } from 'expo-router'
import { MobileAccountLayout } from 'app/features/account/AccountScreenLayout'

export default function AccountScreen() {
  return (
    <>
      <Stack.Screen
        options={{
          title: 'Account',
        }}
      />
      <ScreenContainer>
        <MobileAccountLayout />
      </ScreenContainer>
    </>
  )
}
