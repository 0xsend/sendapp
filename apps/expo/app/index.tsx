import { YStack } from '@my/ui'
import { useQueryClient } from '@tanstack/react-query'
import { AuthCarouselContext } from 'app/features/auth/AuthCarouselContext'
import { HomeScreen } from 'app/features/home/screen'
import { SplashScreen } from 'app/features/splash/screen'
import type { GetPlaiceholderImage } from 'app/utils/getPlaiceholderImage'
import { useSendAccount } from 'app/utils/send-accounts/useSendAccounts'
import { useUser } from 'app/utils/useUser'
import { Stack } from 'expo-router'
import { useCallback, useEffect, useState } from 'react'

// Hardcoded carousel images for mobile - using type assertion to avoid complexities with GetPlaiceholderImage type
const mobileCarouselImages = [
  {
    color: {
      r: 8,
      g: 8,
      b: 8,
      hex: '#080808',
    },
    css: {
      backgroundImage:
        'linear-gradient(90deg, rgb(136,120,61) 25%,rgb(121,163,146) 25% 50%,rgb(77,88,94) 50% 75%,rgb(82,96,95) 75% 100%),linear-gradient(90deg, rgb(105,109,95) 25%,rgb(103,136,144) 25% 50%,rgb(79,59,45) 50% 75%,rgb(92,89,78) 75% 100%),linear-gradient(90deg, rgb(81,71,69) 25%,rgb(96,70,60) 25% 50%,rgb(121,74,27) 50% 75%,rgb(96,64,42) 75% 100%)',
      backgroundPosition: '0 0 ,0 50%,0 100%',
      backgroundSize: '100% 33.333333333333336%',
      backgroundRepeat: 'no-repeat',
    },
    base64:
      'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAADCAIAAAA7ljmRAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAMElEQVR4nGNYPzfh5v/PVi5enmFhDOXVuXP3H5dmYPDxtGPQVJS3UpaKN2ExEmcAADiuDGUiluR6AAAAAElFTkSuQmCC',
    pixels: [
      [
        {
          r: 136,
          g: 120,
          b: 61,
        },
        {
          r: 121,
          g: 163,
          b: 146,
        },
        {
          r: 77,
          g: 88,
          b: 94,
        },
        {
          r: 82,
          g: 96,
          b: 95,
        },
      ],
      [
        {
          r: 105,
          g: 109,
          b: 95,
        },
        {
          r: 103,
          g: 136,
          b: 144,
        },
        {
          r: 79,
          g: 59,
          b: 45,
        },
        {
          r: 92,
          g: 89,
          b: 78,
        },
      ],
      [
        {
          r: 81,
          g: 71,
          b: 69,
        },
        {
          r: 96,
          g: 70,
          b: 60,
        },
        {
          r: 121,
          g: 74,
          b: 27,
        },
        {
          r: 96,
          g: 64,
          b: 42,
        },
      ],
    ],
    svg: [
      'svg',
      {
        xmlns: 'http://www.w3.org/2000/svg',
        width: '100%',
        height: '100%',
        shapeRendering: 'crispEdges',
        preserveAspectRatio: 'none',
        viewBox: '0 0 4 3',
        style: {
          position: 'absolute',
          top: '50%',
          left: '50%',
          transformOrigin: 'top left',
          transform: 'translate(-50%, -50%)',
          right: 0,
          bottom: 0,
        },
      },
      [
        [
          'rect',
          {
            fill: 'rgb(136,120,61)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 0,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(121,163,146)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 1,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(77,88,94)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 2,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(82,96,95)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 3,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(105,109,95)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 0,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(103,136,144)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 1,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(79,59,45)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 2,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(92,89,78)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 3,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(81,71,69)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 0,
            y: 2,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(96,70,60)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 1,
            y: 2,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(121,74,27)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 2,
            y: 2,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(96,64,42)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 3,
            y: 2,
          },
        ],
      ],
    ],
    img: {
      src: 'https://ghassets.send.app/app_images/auth_image_3.jpg',
      height: 1152,
      width: 1440,
    },
  },
  {
    color: {
      r: 8,
      g: 8,
      b: 8,
      hex: '#080808',
    },
    css: {
      backgroundImage:
        'linear-gradient(90deg, rgb(104,149,126) 25%,rgb(0,145,115) 25% 50%,rgb(133,92,59) 50% 75%,rgb(98,106,87) 75% 100%),linear-gradient(90deg, rgb(66,110,73) 25%,rgb(0,153,62) 25% 50%,rgb(14,83,38) 50% 75%,rgb(50,72,57) 75% 100%),linear-gradient(90deg, rgb(0,101,50) 25%,rgb(0,164,42) 25% 50%,rgb(0,159,24) 50% 75%,rgb(0,75,32) 75% 100%)',
      backgroundPosition: '0 0 ,0 50%,0 100%',
      backgroundSize: '100% 33.333333333333336%',
      backgroundRepeat: 'no-repeat',
    },
    base64:
      'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAADCAIAAAA7ljmRAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAMElEQVR4nGO48P954aMD03K8yutzGAJqw+q+TWZQY2AQZWBgCFOd9X9a3/96BlEGAGSjD6iQjVIFAAAAAElFTkSuQmCC',
    pixels: [
      [
        {
          r: 104,
          g: 149,
          b: 126,
        },
        {
          r: 0,
          g: 145,
          b: 115,
        },
        {
          r: 133,
          g: 92,
          b: 59,
        },
        {
          r: 98,
          g: 106,
          b: 87,
        },
      ],
      [
        {
          r: 66,
          g: 110,
          b: 73,
        },
        {
          r: 0,
          g: 153,
          b: 62,
        },
        {
          r: 14,
          g: 83,
          b: 38,
        },
        {
          r: 50,
          g: 72,
          b: 57,
        },
      ],
      [
        {
          r: 0,
          g: 101,
          b: 50,
        },
        {
          r: 0,
          g: 164,
          b: 42,
        },
        {
          r: 0,
          g: 159,
          b: 24,
        },
        {
          r: 0,
          g: 75,
          b: 32,
        },
      ],
    ],
    svg: [
      'svg',
      {
        xmlns: 'http://www.w3.org/2000/svg',
        width: '100%',
        height: '100%',
        shapeRendering: 'crispEdges',
        preserveAspectRatio: 'none',
        viewBox: '0 0 4 3',
        style: {
          position: 'absolute',
          top: '50%',
          left: '50%',
          transformOrigin: 'top left',
          transform: 'translate(-50%, -50%)',
          right: 0,
          bottom: 0,
        },
      },
      [
        [
          'rect',
          {
            fill: 'rgb(104,149,126)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 0,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(0,145,115)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 1,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(133,92,59)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 2,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(98,106,87)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 3,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(66,110,73)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 0,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(0,153,62)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 1,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(14,83,38)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 2,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(50,72,57)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 3,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(0,101,50)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 0,
            y: 2,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(0,164,42)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 1,
            y: 2,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(0,159,24)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 2,
            y: 2,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(0,75,32)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 3,
            y: 2,
          },
        ],
      ],
    ],
    img: {
      src: 'https://ghassets.send.app/app_images/auth_image_2.jpg',
      height: 1152,
      width: 1440,
    },
  },
  {
    color: {
      r: 8,
      g: 8,
      b: 24,
      hex: '#080818',
    },
    css: {
      backgroundImage:
        'linear-gradient(90deg, rgb(85,116,122) 25%,rgb(130,156,155) 25% 50%,rgb(104,115,117) 50% 75%,rgb(95,109,109) 75% 100%),linear-gradient(90deg, rgb(91,113,118) 25%,rgb(106,137,140) 25% 50%,rgb(122,120,115) 50% 75%,rgb(43,47,49) 75% 100%),linear-gradient(90deg, rgb(64,76,77) 25%,rgb(57,75,80) 25% 50%,rgb(47,48,46) 50% 75%,rgb(21,23,26) 75% 100%)',
      backgroundPosition: '0 0 ,0 50%,0 100%',
      backgroundSize: '100% 33.333333333333336%',
      backgroundRepeat: 'no-repeat',
    },
    base64:
      'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAADCAIAAAA7ljmRAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAMklEQVR4nAEnANj/AJO0uub//6i0tpmoqACXr7S73N/CwLsxNTcAXWlqVGdsNjc1AQUKz7ATNB5aY3kAAAAASUVORK5CYII=',
    pixels: [
      [
        {
          r: 85,
          g: 116,
          b: 122,
        },
        {
          r: 130,
          g: 156,
          b: 155,
        },
        {
          r: 104,
          g: 115,
          b: 117,
        },
        {
          r: 95,
          g: 109,
          b: 109,
        },
      ],
      [
        {
          r: 91,
          g: 113,
          b: 118,
        },
        {
          r: 106,
          g: 137,
          b: 140,
        },
        {
          r: 122,
          g: 120,
          b: 115,
        },
        {
          r: 43,
          g: 47,
          b: 49,
        },
      ],
      [
        {
          r: 64,
          g: 76,
          b: 77,
        },
        {
          r: 57,
          g: 75,
          b: 80,
        },
        {
          r: 47,
          g: 48,
          b: 46,
        },
        {
          r: 21,
          g: 23,
          b: 26,
        },
      ],
    ],
    svg: [
      'svg',
      {
        xmlns: 'http://www.w3.org/2000/svg',
        width: '100%',
        height: '100%',
        shapeRendering: 'crispEdges',
        preserveAspectRatio: 'none',
        viewBox: '0 0 4 3',
        style: {
          position: 'absolute',
          top: '50%',
          left: '50%',
          transformOrigin: 'top left',
          transform: 'translate(-50%, -50%)',
          right: 0,
          bottom: 0,
        },
      },
      [
        [
          'rect',
          {
            fill: 'rgb(85,116,122)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 0,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(130,156,155)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 1,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(104,115,117)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 2,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(95,109,109)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 3,
            y: 0,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(91,113,118)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 0,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(106,137,140)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 1,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(122,120,115)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 2,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(43,47,49)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 3,
            y: 1,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(64,76,77)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 0,
            y: 2,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(57,75,80)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 1,
            y: 2,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(47,48,46)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 2,
            y: 2,
          },
        ],
        [
          'rect',
          {
            fill: 'rgb(21,23,26)',
            // @ts-expect-error - still works
            'fill-opacity': 1,
            width: 1,
            height: 1,
            x: 3,
            y: 2,
          },
        ],
      ],
    ],
    img: {
      src: 'https://ghassets.send.app/app_images/auth_image_1.jpg',
      height: 1152,
      width: 1440,
    },
  },
] satisfies GetPlaiceholderImage[]

export default function Screen() {
  const { session } = useUser()
  const [carouselImages, setCarouselImages] = useState<GetPlaiceholderImage[]>([])
  const [carouselProgress, setCarouselProgress] = useState(0)
  const queryClient = useQueryClient()

  const cancelAndRemoveAccountsQueries = useCallback(async () => {
    if (!session) {
      const options = { queryKey: [useSendAccount.queryKey] }
      await queryClient.cancelQueries(options)
      queryClient.removeQueries(options)
    }
  }, [session, queryClient])

  useEffect(() => {
    // @ts-expect-error - still works
    if (carouselImages.length === 0) setCarouselImages(mobileCarouselImages)
  }, [carouselImages])

  useEffect(() => {
    void cancelAndRemoveAccountsQueries()
  }, [cancelAndRemoveAccountsQueries])

  return (
    <>
      <Stack.Screen
        options={{
          title: 'Home',
          headerShown: false,
        }}
      />

      {session ? (
        <YStack f={1}>
          <HomeScreen />
        </YStack>
      ) : (
        <AuthCarouselContext.Provider
          value={{
            carouselImages,
            setCarouselImages,
            carouselProgress,
            setCarouselProgress,
          }}
        >
          <SplashScreen />
        </AuthCarouselContext.Provider>
      )}
    </>
  )
}
