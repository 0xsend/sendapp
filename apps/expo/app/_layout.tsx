import type { Session } from '@supabase/supabase-js'
import { Provider, loadThemePromise } from 'app/provider'
import { supabase } from 'app/utils/supabase/client.native'
import { SplashScreen, Stack } from 'expo-router'
import { useCallback, useEffect, useState } from 'react'
import { LogBox, View } from 'react-native'
import { GestureHandlerRootView } from 'react-native-gesture-handler'

SplashScreen.preventAutoHideAsync()

LogBox.ignoreLogs([
  'Cannot update a component',
  'You are setting the style',
  'No route',
  'duplicate ID',
  'Require cycle',
])

export default function HomeLayout() {
  // TODO: use fonts
  // const [fontLoaded] = useFonts({
  //   'DM Sans': require('@tamagui/font-dm-sans/fonts/static/DMSans-Medium.ttf'),
  //   'DM Sans Bold': require('@tamagui/font-dm-sans/fonts/static/DMSans-Bold.ttf'),
  // })
  const [fontLoaded] = useState(true)

  const [themeLoaded, setThemeLoaded] = useState(false)
  const [sessionLoadAttempted, setSessionLoadAttempted] = useState(false)
  const [initialSession, setInitialSession] = useState<Session | null>(null)
  useEffect(() => {
    supabase.auth
      .getSession()
      .then(({ data }) => {
        if (data) {
          setInitialSession(data.session)
        }
      })
      .finally(() => {
        setSessionLoadAttempted(true)
      })
  }, [])

  useEffect(() => {
    loadThemePromise.then(() => {
      setThemeLoaded(true)
    })
  }, [])

  const onLayoutRootView = useCallback(async () => {
    if (fontLoaded && sessionLoadAttempted) {
      await SplashScreen.hideAsync()
    }
  }, [fontLoaded, sessionLoadAttempted])

  if (!themeLoaded || !fontLoaded || !sessionLoadAttempted) {
    return null
  }

  return (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <View style={{ flex: 1 }} onLayout={onLayoutRootView}>
        <Provider initialSession={initialSession}>
          <Stack screenOptions={{ headerShown: false }}>
            <Stack.Screen name="index" options={{ headerShown: false }} />
            <Stack.Screen name="(tabs)" options={{ headerShown: false }} />
            <Stack.Screen name="(auth)" options={{ headerShown: false }} />
            <Stack.Screen name="send" options={{ headerShown: false }} />
            <Stack.Screen name="deposit" options={{ headerShown: false }} />
            <Stack.Screen name="trade" options={{ headerShown: false }} />
            <Stack.Screen name="sendpot" options={{ headerShown: false }} />
            <Stack.Screen name="leaderboard" options={{ headerShown: false }} />
            <Stack.Screen name="secret-shop" options={{ headerShown: false }} />
            <Stack.Screen
              name="settings/index"
              options={{
                headerShown: true,
                headerBackTitle: 'Back',
              }}
            />
          </Stack>
        </Provider>
      </View>
    </GestureHandlerRootView>
  )
}
