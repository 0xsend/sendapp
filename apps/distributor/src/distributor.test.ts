// @ts-expect-error set __DEV__ for code shared between server and client
globalThis.__DEV__ = true

import request from 'supertest'
import { describe, expect, it } from 'bun:test'
import app from './app'

describe('Root Route', () => {
  it('should return correct response for the root route', async () => {
    const res = await request(app).get('/')

    expect(res.statusCode).toBe(200)
    expect(res.body).toEqual({ root: true })
  })
})

describe('Distributor Route', () => {
  it('should reject unauthorized requests', async () => {
    const res = await request(app).post('/distributor')

    expect(res.statusCode).toBe(401)
    expect(res.body).toEqual('Unauthorized')
  })

  it('should handle authorization correctly', async () => {
    const res = await request(app).get('/distributor')

    expect(res.statusCode).toBe(200)
    expect(res.body).toMatchObject({
      distributor: true,
      running: true,
    })
  })

  it('should perform distributor logic correctly', async () => {
    // get latest distribution id from API
    let lastDistributionId: number
    while (true) {
      const res = await request(app).get('/distributor')
      expect(res.statusCode).toBe(200)
      if (res.body.lastDistributionId) {
        lastDistributionId = res.body.lastDistributionId
        break
      }
      await new Promise((resolve) => setTimeout(resolve, 1000))
    }

    expect(lastDistributionId).toBeDefined()
  }, 10_000)
})
