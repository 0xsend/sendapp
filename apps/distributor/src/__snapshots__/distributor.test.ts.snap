// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Distributor Route should return a merkle root: distribution 4 merkle root 1`] = `
{
  "root": "0x0323baed48d30414af5f66d611d02e0c905bf97ca10b60942104a85b86cab876",
  "total": 627736455,
}
`;
